File: root_VS2013_TemplateBase/programs/C#/Samples/2CS_sample/AsyncEvent_sample/WindowsFormsApplication/Form1.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Windows.Forms;

using System.IO;
using System.Threading;
using System.Diagnostics;

using Touryo.Infrastructure.Public.IO;
using Touryo.Infrastructure.Framework.RichClient.Asynchronous;

namespace WindowsFormsApplication
{
    public partial class Form1 : Form
    {
        #region メンバ変数

        /// <summary>名前付きパイプ・サーバ名</summary>
        private string NPS = null;
        /// <summary>名前付きパイプ・クライアント</summary>
        private string[] NPCS = null;

        /// <summary>登録エントリ（Thread）</summary>
        private AsyncEventEntry AeeTh = null;
        /// <summary>登録エントリ（ThreadPool）</summary>
        private AsyncEventEntry AeePl = null;
        /// <summary>登録エントリ（WinForm）</summary>
        private AsyncEventEntry AeeWin = null;
        /// <summary>登録エントリ（WPF）</summary>
        private AsyncEventEntry AeeWPF = null;

        #endregion

        #region 開始・終了処理

        #region 開始処理

        /// <summary>コンストラクタ</summary>
        public Form1()
        {
            InitializeComponent();

            // サーバを起動
            string[] args = Environment.CommandLine.Split('/');
            args = args[1].Trim().Split(',');

            // this.NPS
            this.NPS = args[0].Trim();
            this.Text = this.NPS;

            // this.NPCS
            this.NPCS = new string[args.Length - 1];
            for (int i = 1; i < args.Length; i++)
            {
                this.NPCS[i-1] = args[i].Trim();
            }

            // 初期化

            // 初めが自分の名称、

            //// ２つ目からが相手の名称
            //MessageBox.Show(
            //    "this.NPS:" + this.NPS
            //    + "\r\nthis.NPCS:" + string.Join(",", this.NPCS),
            //    "コマンドラインのチェック");

            AsyncEventFx.Init(this.NPS, this.NPCS, 3000);
        }

        /// <summary>初期処理</summary>
        private void Form1_Load(object sender, EventArgs e)
        {
            // 登録エントリ

            // スレッド
            this.AeeTh = new AsyncEventEntry(
                AsyncEventEnum.EventClass.Thread, "Thread", null,
                new ParameterizedThreadStart(this.ParameterizedThreadStartDgt));

            AsyncEventFx.RegisterAsyncEvent(this.AeeTh);

            // ---

            // スレッド プール
            this.AeePl = new AsyncEventEntry(
                AsyncEventEnum.EventClass.ThreadPool, "ThreadPool", null,
                new WaitCallback(this.WaitCallbackDgt));

            AsyncEventFx.RegisterAsyncEvent(this.AeePl);

            // ---

            // WinForm
            this.AeeWin = new AsyncEventEntry(
                AsyncEventEnum.EventClass.WinForm, "WinForm", this,
                new AsyncEventFx.SetResultDelegate(this.SetResultDgt));

            AsyncEventFx.RegisterAsyncEvent(this.AeeWin);
            
            // ---

            // WPF
            this.AeeWPF = new AsyncEventEntry(
                AsyncEventEnum.EventClass.WPF, "WPF", this,
                new AsyncEventFx.SetResultDelegate(this.SetResultDgt));

            AsyncEventFx.RegisterAsyncEvent(this.AeeWPF);
        }
        
        #endregion

        #region 終了処理

        /// <summary>終了処理</summary>
        private void Form1_FormClosed(object sender, FormClosedEventArgs e)
        {
            // 終了
            AsyncEventFx.Final();
        }

        #endregion

        #endregion

        #region 各種デリゲード

        /// <summary>デリゲード</summary>
        private void ParameterizedThreadStartDgt(object obj)
        {
            object[] param = (object[])obj;
            AsyncEventHeader aeh = (AsyncEventHeader)param[0];
            string msg = (string)BinarySerialize.BytesToObject((byte[])param[1]);

            // ファイルにテキストを書き出し。
            using (StreamWriter sw = new StreamWriter(this.NPS + @"_test_pts.txt", true))
            {
                sw.WriteLine("\r\n" + this.NPS + " - "
                + DateTime.Now.ToString("yyyy/MM/dd HH:mm:ss")
                + "WindowsFormsApplication.Form1.ParameterizedThreadStartDgtメソッドが呼ばれた。"
                + "\r\nDstEventClass:" + aeh.DstEventClass.ToString()
                + "\r\nDstFuncID:" + (new string(aeh.DstFuncID)).Trim()
                + "\r\nSrcEventClass:" + aeh.SrcEventClass.ToString()
                + "\r\nSrcFuncID:" + (new string(aeh.SrcFuncID)).Trim()
                + "\r\nSrcPipeName:" + (new string(aeh.SrcPipeName)).Trim()
                + "\r\nメッセージ:" + msg, this.NPS);
            }
        }

        /// <summary>デリゲード</summary>
        private void WaitCallbackDgt(object state)
        {
            object[] param = (object[])state;
            AsyncEventHeader aeh = (AsyncEventHeader)param[0];
            string msg = (string)BinarySerialize.BytesToObject((byte[])param[1]);

            // ファイルにテキストを書き出し。
            using (StreamWriter sw = new StreamWriter(this.NPS + @"_test_tpl.txt", true))
            {
                sw.WriteLine("\r\n" + this.NPS + " - "
                + DateTime.Now.ToString("yyyy/MM/dd HH:mm:ss")
                + "WindowsFormsApplication.Form1.WaitCallbackDgtメソッドが呼ばれた。"
                + "\r\nDstEventClass:" + aeh.DstEventClass.ToString()
                + "\r\nDstFuncID:" + (new string(aeh.DstFuncID)).Trim()
                + "\r\nSrcEventClass:" + aeh.SrcEventClass.ToString()
                + "\r\nSrcFuncID:" + (new string(aeh.SrcFuncID)).Trim()
                + "\r\nSrcPipeName:" + (new string(aeh.SrcPipeName)).Trim()
                + "\r\nメッセージ:" + msg, this.NPS);
            }
        }

        /// <summary>デリゲード</summary>
        private void SetResultDgt(object result)
        {
            object[] param = (object[])result;
            AsyncEventHeader aeh = (AsyncEventHeader)param[0];
            string msg = (string)BinarySerialize.BytesToObject((byte[])param[1]);

            MessageBox.Show(this.NPS + " - " 
                + DateTime.Now.ToString("yyyy/MM/dd HH:mm:ss")
                + "WindowsFormsApplication.Form1.SetResultDgtメソッドが呼ばれた。"
                + "\r\nDstEventClass:" + aeh.DstEventClass.ToString()
                + "\r\nDstFuncID:" + (new string(aeh.DstFuncID)).Trim()
                + "\r\nSrcEventClass:" + aeh.SrcEventClass.ToString()
                + "\r\nSrcFuncID:" + (new string(aeh.SrcFuncID)).Trim()
                + "\r\nSrcPipeName:" + (new string(aeh.SrcPipeName)).Trim()
                + "\r\nメッセージ:" + msg, this.NPS);
        }

        #endregion

        #region 各種ボタン（イベント送信）

        /// <summary>WinFormのThread</summary>
        private void button1_Click(object sender, EventArgs e)
        {
            byte[] bytes = BinarySerialize.ObjectToBytes(this.txtMSG.Text);

            AsyncEventFx.SendAsyncEvent(
                AsyncEventEnum.EventClass.Thread, "Thread",
                AsyncEventEnum.EventClass.Thread, "Thread",
                this.NPCS[0], this.NPS, (uint)bytes.Length, bytes);
        }

        /// <summary>WinFormのThreadPool</summary>
        private void button2_Click(object sender, EventArgs e)
        {
            byte[] bytes = BinarySerialize.ObjectToBytes(this.txtMSG.Text);

            AsyncEventFx.SendAsyncEvent(
                AsyncEventEnum.EventClass.ThreadPool, "ThreadPool",
                AsyncEventEnum.EventClass.ThreadPool, "ThreadPool",
                this.NPCS[0], this.NPS, (uint)bytes.Length, bytes);
        }

        /// <summary>WinFormのUIInvoke</summary>
        private void button3_Click(object sender, EventArgs e)
        {
            byte[] bytes = BinarySerialize.ObjectToBytes(this.txtMSG.Text);

            AsyncEventFx.SendAsyncEvent(
                AsyncEventEnum.EventClass.WinForm, "WinForm",
                AsyncEventEnum.EventClass.WinForm, "WinForm",
                this.NPCS[0], this.NPS, (uint)bytes.Length, bytes);
        }

        /// <summary>WPFのUIInvoke</summary>
        private void button4_Click(object sender, EventArgs e)
        {
            byte[] bytes = BinarySerialize.ObjectToBytes(this.txtMSG.Text);

            AsyncEventFx.SendAsyncEvent(
                AsyncEventEnum.EventClass.WPF, "WPF",
                AsyncEventEnum.EventClass.WPF, "WPF",
                this.NPCS[1], this.NPS, (uint)bytes.Length, bytes);
        }

        /// <summary>へんなところ</summary>
        private void button5_Click(object sender, EventArgs e)
        {
            byte[] bytes = BinarySerialize.ObjectToBytes(this.txtMSG.Text);

            AsyncEventFx.SendAsyncEvent(
                AsyncEventEnum.EventClass.WPF, "いいい",
                AsyncEventEnum.EventClass.WPF, "いいい",
                this.NPCS[1], "あああ", (uint)bytes.Length, bytes);
            //this.NPCS[1], this.NPS, (uint)bytes.Length, bytes);
        }

        #endregion

        #region 各種ボタン（エントリ）

        /// <summary>エントリを登録</summary>
        private void button6_Click(object sender, EventArgs e)
        {
            AsyncEventFx.RegisterAsyncEvent(this.AeeTh);
            AsyncEventFx.RegisterAsyncEvent(this.AeePl);
            AsyncEventFx.RegisterAsyncEvent(this.AeeWin);
            AsyncEventFx.RegisterAsyncEvent(this.AeeWPF);
        }

        /// <summary>エントリを削除</summary>
        private void button7_Click(object sender, EventArgs e)
        {
            AsyncEventFx.UnRegisterAsyncEvent(this.AeeTh);
            AsyncEventFx.UnRegisterAsyncEvent(this.AeePl);
            AsyncEventFx.UnRegisterAsyncEvent(this.AeeWin);
            AsyncEventFx.UnRegisterAsyncEvent(this.AeeWPF);
        }

        #endregion
    }
}

File: root_VS2013_TemplateBase/programs/VB/Samples/2CS_sample/AsyncEvent_sample/WindowsFormsApplication/Form1.vb
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Linq
Imports System.Text
Imports System.Windows.Forms

Imports System.IO
Imports System.Threading
Imports System.Diagnostics

Imports Touryo.Infrastructure.Public.IO
Imports Touryo.Infrastructure.Framework.RichClient.Asynchronous

Partial Public Class Form1
#Region "メンバ変数"

    ''' <summary>名前付きパイプ・サーバ名</summary>
    Private NPS As String = Nothing
    ''' <summary>名前付きパイプ・クライアント</summary>
    Private NPCS As String() = Nothing

    ''' <summary>登録エントリ（Thread）</summary>
    Private AeeTh As AsyncEventEntry = Nothing
    ''' <summary>登録エントリ（ThreadPool）</summary>
    Private AeePl As AsyncEventEntry = Nothing
    ''' <summary>登録エントリ（WinForm）</summary>
    Private AeeWin As AsyncEventEntry = Nothing
    ''' <summary>登録エントリ（WPF）</summary>
    Private AeeWPF As AsyncEventEntry = Nothing

#End Region

#Region "開始・終了処理"

#Region "開始処理"

    ''' <summary>コンストラクタ</summary>
    Public Sub New()
        InitializeComponent()

        ' サーバを起動
        Dim args As String() = Environment.CommandLine.Split("/"c)
        args = args(1).Trim().Split(","c)

        ' this.NPS
        Me.NPS = args(0).Trim()
        Me.Text = Me.NPS

        ' this.NPCS
        Me.NPCS = New String(args.Length - 2) {}
        For i As Integer = 1 To args.Length - 1
            Me.NPCS(i - 1) = args(i).Trim()
        Next

        ' 初期化

        ' 初めが自分の名称、

        '/ ２つ目からが相手の名称
        'MessageBox.Show(
        '    "this.NPS:" + this.NPS
        '    + "\r\nthis.NPCS:" + string.Join(",", this.NPCS),
        '    "コマンドラインのチェック");

        AsyncEventFx.Init(Me.NPS, Me.NPCS, 3000)
    End Sub

    ''' <summary>初期処理</summary>
    Private Sub Form1_Load(ByVal sender As Object, ByVal e As EventArgs) Handles MyBase.Load
        ' 登録エントリ

        ' スレッド
        Me.AeeTh = New AsyncEventEntry(AsyncEventEnum.EventClass.Thread, "Thread", Nothing, New ParameterizedThreadStart(AddressOf Me.ParameterizedThreadStartDgt))
        AsyncEventFx.RegisterAsyncEvent(Me.AeeTh)

        ' ---

        ' スレッド プール
        Me.AeePl = New AsyncEventEntry(AsyncEventEnum.EventClass.ThreadPool, "ThreadPool", Nothing, New WaitCallback(AddressOf Me.WaitCallbackDgt))
        AsyncEventFx.RegisterAsyncEvent(Me.AeePl)

        ' ---

        ' WinForm
        Me.AeeWin = New AsyncEventEntry(AsyncEventEnum.EventClass.WinForm, "WinForm", Me, New AsyncEventFx.SetResultDelegate(AddressOf Me.SetResultDgt))
        AsyncEventFx.RegisterAsyncEvent(Me.AeeWin)

        ' ---

        ' WPF
        Me.AeeWPF = New AsyncEventEntry(AsyncEventEnum.EventClass.WPF, "WPF", Me, New AsyncEventFx.SetResultDelegate(AddressOf Me.SetResultDgt))
        AsyncEventFx.RegisterAsyncEvent(Me.AeeWPF)
    End Sub

#End Region

#Region "終了処理"

    ''' <summary>終了処理</summary>
    Private Sub Form1_FormClosed(ByVal sender As Object, ByVal e As FormClosedEventArgs) Handles MyBase.FormClosed
        ' 終了
        AsyncEventFx.Final()
    End Sub

#End Region

#End Region

#Region "各種デリゲード"

    ''' <summary>デリゲード</summary>
    Private Sub ParameterizedThreadStartDgt(ByVal obj As Object)
        Dim param As Object() = DirectCast(obj, Object())
        Dim aeh As AsyncEventHeader = DirectCast(param(0), AsyncEventHeader)
        Dim msg As String = DirectCast(BinarySerialize.BytesToObject(DirectCast(param(1), Byte())), String)

        ' ファイルにテキストを書き出し。
        Using sw As New StreamWriter(Me.NPS & "_test_pts.txt", True)
            sw.WriteLine(vbCr & vbLf & Me.NPS & " - " & DateTime.Now.ToString("yyyy/MM/dd HH:mm:ss") & "WindowsFormsApplication.Form1.ParameterizedThreadStartDgtメソッドが呼ばれた。" & vbCr & vbLf & "DstEventClass:" & aeh.DstEventClass.ToString() & vbCr & vbLf & "DstFuncID:" & (New String(aeh.DstFuncID)).Trim() & vbCr & vbLf & "SrcEventClass:" & aeh.SrcEventClass.ToString() & vbCr & vbLf & "SrcFuncID:" & (New String(aeh.SrcFuncID)).Trim() & vbCr & vbLf & "SrcPipeName:" & (New String(aeh.SrcPipeName)).Trim() & vbCr & vbLf & "メッセージ:" & msg, Me.NPS)
        End Using
    End Sub

    ''' <summary>デリゲード</summary>
    Private Sub WaitCallbackDgt(ByVal state As Object)
        Dim param As Object() = DirectCast(state, Object())
        Dim aeh As AsyncEventHeader = DirectCast(param(0), AsyncEventHeader)
        Dim msg As String = DirectCast(BinarySerialize.BytesToObject(DirectCast(param(1), Byte())), String)

        ' ファイルにテキストを書き出し。
        Using sw As New StreamWriter(Me.NPS & "_test_tpl.txt", True)
            sw.WriteLine(vbCr & vbLf & Me.NPS & " - " & DateTime.Now.ToString("yyyy/MM/dd HH:mm:ss") & "WindowsFormsApplication.Form1.WaitCallbackDgtメソッドが呼ばれた。" & vbCr & vbLf & "DstEventClass:" & aeh.DstEventClass.ToString() & vbCr & vbLf & "DstFuncID:" & (New String(aeh.DstFuncID)).Trim() & vbCr & vbLf & "SrcEventClass:" & aeh.SrcEventClass.ToString() & vbCr & vbLf & "SrcFuncID:" & (New String(aeh.SrcFuncID)).Trim() & vbCr & vbLf & "SrcPipeName:" & (New String(aeh.SrcPipeName)).Trim() & vbCr & vbLf & "メッセージ:" & msg, Me.NPS)
        End Using
    End Sub

    ''' <summary>デリゲード</summary>
    Private Sub SetResultDgt(ByVal result As Object)
        Dim param As Object() = DirectCast(result, Object())
        Dim aeh As AsyncEventHeader = DirectCast(param(0), AsyncEventHeader)
        Dim msg As String = DirectCast(BinarySerialize.BytesToObject(DirectCast(param(1), Byte())), String)

        MessageBox.Show(Me.NPS & " - " & DateTime.Now.ToString("yyyy/MM/dd HH:mm:ss") & "WindowsFormsApplication.Form1.SetResultDgtメソッドが呼ばれた。" & vbCr & vbLf & "DstEventClass:" & aeh.DstEventClass.ToString() & vbCr & vbLf & "DstFuncID:" & (New String(aeh.DstFuncID)).Trim() & vbCr & vbLf & "SrcEventClass:" & aeh.SrcEventClass.ToString() & vbCr & vbLf & "SrcFuncID:" & (New String(aeh.SrcFuncID)).Trim() & vbCr & vbLf & "SrcPipeName:" & (New String(aeh.SrcPipeName)).Trim() & vbCr & vbLf & "メッセージ:" & msg, Me.NPS)
    End Sub

#End Region

#Region "各種ボタン（イベント送信）"

    ''' <summary>WinFormのThread</summary>
    Private Sub button1_Click(ByVal sender As Object, ByVal e As EventArgs) Handles button1.Click
        Dim bytes As Byte() = BinarySerialize.ObjectToBytes(Me.txtMSG.Text)

        AsyncEventFx.SendAsyncEvent(AsyncEventEnum.EventClass.Thread, "Thread", AsyncEventEnum.EventClass.Thread, "Thread", Me.NPCS(0), Me.NPS, _
         CUInt(bytes.Length), bytes)
    End Sub

    ''' <summary>WinFormのThreadPool</summary>
    Private Sub button2_Click(ByVal sender As Object, ByVal e As EventArgs) Handles button2.Click
        Dim bytes As Byte() = BinarySerialize.ObjectToBytes(Me.txtMSG.Text)

        AsyncEventFx.SendAsyncEvent(AsyncEventEnum.EventClass.ThreadPool, "ThreadPool", AsyncEventEnum.EventClass.ThreadPool, "ThreadPool", Me.NPCS(0), Me.NPS, _
         CUInt(bytes.Length), bytes)
    End Sub

    ''' <summary>WinFormのUIInvoke</summary>
    Private Sub button3_Click(ByVal sender As Object, ByVal e As EventArgs) Handles button3.Click
        Dim bytes As Byte() = BinarySerialize.ObjectToBytes(Me.txtMSG.Text)

        AsyncEventFx.SendAsyncEvent(AsyncEventEnum.EventClass.WinForm, "WinForm", AsyncEventEnum.EventClass.WinForm, "WinForm", Me.NPCS(0), Me.NPS, _
         CUInt(bytes.Length), bytes)
    End Sub

    ''' <summary>WPFのUIInvoke</summary>
    Private Sub button4_Click(ByVal sender As Object, ByVal e As EventArgs) Handles button4.Click
        Dim bytes As Byte() = BinarySerialize.ObjectToBytes(Me.txtMSG.Text)

        AsyncEventFx.SendAsyncEvent(AsyncEventEnum.EventClass.WPF, "WPF", AsyncEventEnum.EventClass.WPF, "WPF", Me.NPCS(1), Me.NPS, _
         CUInt(bytes.Length), bytes)
    End Sub

    ''' <summary>へんなところ</summary>
    Private Sub button5_Click(ByVal sender As Object, ByVal e As EventArgs) Handles button5.Click
        Dim bytes As Byte() = BinarySerialize.ObjectToBytes(Me.txtMSG.Text)

        AsyncEventFx.SendAsyncEvent(AsyncEventEnum.EventClass.WPF, "いいい", AsyncEventEnum.EventClass.WPF, "いいい", Me.NPCS(1), "あああ", _
         CUInt(bytes.Length), bytes)
        'this.NPCS[1], this.NPS, (uint)bytes.Length, bytes);
    End Sub

#End Region

#Region "各種ボタン（エントリ）"

    ''' <summary>エントリを登録</summary>
    Private Sub button6_Click(ByVal sender As Object, ByVal e As EventArgs) Handles button6.Click
        AsyncEventFx.RegisterAsyncEvent(Me.AeeTh)
        AsyncEventFx.RegisterAsyncEvent(Me.AeePl)
        AsyncEventFx.RegisterAsyncEvent(Me.AeeWin)
        AsyncEventFx.RegisterAsyncEvent(Me.AeeWPF)
    End Sub

    ''' <summary>エントリを削除</summary>
    Private Sub button7_Click(ByVal sender As Object, ByVal e As EventArgs) Handles button7.Click
        AsyncEventFx.UnRegisterAsyncEvent(Me.AeeTh)
        AsyncEventFx.UnRegisterAsyncEvent(Me.AeePl)
        AsyncEventFx.UnRegisterAsyncEvent(Me.AeeWin)
        AsyncEventFx.UnRegisterAsyncEvent(Me.AeeWPF)
    End Sub

#End Region
End Class
